### PR TITLE
fix(plans): preserve revision scope and allow free-form answers

### PR DESCRIPTION
**Pull request by zjw49246**
## Summary

- preserve the original Plan request, Run semantics, Base Version review state, current candidate, prior Reviewer feedback, and initial/current Run attachments across revision rounds
- make Reviewers compare revisions against the Base Version and close prior feedback without treating unchanged original requirements as out of scope
- allow required choice questions to be answered solely through non-empty Additional context, with a UI action that clears an unsuitable radio choice
- expose a dedicated revision action when Reviewer feedback remains after the review-round limit is exhausted

## Validation

- `uv run --frozen pytest -q backend/tests/test_plan_agent_runner.py backend/tests/test_plan_resources.py`: 103 passed; one pre-existing SQLite concurrency test fails identically on `origin/main`
- `npx vitest run src/components/PlanReview`: 48 passed
- targeted PlanReview ESLint: passed
- `git diff --check`: passed
- production build reaches TypeScript validation but is currently blocked by the existing `origin/main` error `TaskForm.tsx(168,62): Property 'session_id' does not exist on type 'Task'`; the affected files are byte-identical to `origin/main`

## Manual checks

- revise only part of a multi-requirement Plan and confirm the untouched original requirements remain in scope
- force multiple Reviewer rounds and confirm prior feedback is explicitly closed
- exhaust review rounds and use the dedicated latest-feedback revision action
- clear a required single-choice answer and submit only Additional context

### CODE REVIEW

**Review by youchengsong — COMMENTED** [review]
# CCM reviewer panel: changes required

The panel found 3 open blocking findings on this reviewed PR head. Address the inline finding threads before merging.

## Reviewer summaries

### Principal engineer — Changes required

The revision-context design is cohesive, but the free-form escape hatch is applied more broadly than the stated required-choice contract.

Finding count: 1 total (1 blocking, 0 advisory).

### Senior engineer — Changes required

Free-form context incorrectly satisfies every required question, including non-choice questions.

Finding count: 1 total (1 blocking, 0 advisory).

### QA engineer — Changes required

The free-form escape hatch is broader than claimed and bypasses every required question type, not only choice questions.

Finding count: 1 total (1 blocking, 0 advisory).

Detailed evidence, impact, required fixes, and verification steps are published in dedicated finding threads/comments.

Reviewed commit: `81db73806048ccb6a812f634f73a56a13f5d8075`.

**Comment by youchengsong** [comment]
**[MEDIUM] Additional context bypasses every required question type**

Reviewer: `principal_engineer` · Category: `correctness`

Evidence: validate_input_answers skips the required-answer error whenever any non-empty response_text exists, without checking question.response_type. The frontend's missingRequired calculation mirrors this global bypass. The PR contract and documentation limit this escape hatch to required choice questions whose generated options do not fit.

Impact: A user can submit arbitrary Additional context while leaving required non-choice questions unanswered; in a mixed questionnaire, one explanation can also silently waive unrelated required fields. The resumed agent then receives null structured answers despite those decisions being declared necessary.

Required fix: Permit the response_text substitution only for the intended choice response types, and apply the same rule in PlanInputForm so client and server validation agree. Other required questions must still contain their own answer.

Verification: Add a request containing both a required choice question and a required non-choice question. Submit null values plus non-empty response_text; assert the server returns 422 for the non-choice question and the UI remains non-submittable until it is answered. Retain the existing choice-only success case.

Finding fingerprint: `d2935c37003264a739d053d133b68fa3c404683581ed5e33c045fae83346c93e`

**Comment by youchengsong** [comment]
**[MEDIUM] Additional context bypasses all required questions**

Reviewer: `qa_engineer` · Category: `correctness`

Evidence: `validate_input_answers` suppresses the required-answer error whenever `has_free_form_answer` is true, without checking `question.response_type`. The matching `PlanInputForm` logic likewise treats any non-empty `additional` value as satisfying every missing required question. Thus required text, numeric, or other non-choice questions can be submitted with null values. Added tests exercise only a required `single_choice`, so this regression is not detected.

Impact: Input requests can be marked answered and the run resumed despite missing mandatory structured answers unrelated to the additional context, violating the claimed choice-only behavior and potentially producing plans from incomplete requirements.

Required fix: Apply the free-form substitution only to the explicitly supported choice response types; continue rejecting missing required non-choice answers. Mirror the same rule in `PlanInputForm` so client and server validation agree.

Verification: Add backend and frontend tests containing both a required single-choice question and a required non-choice question: additional context may replace the choice, but submission must remain blocked/rejected until the non-choice question is answered.

Finding fingerprint: `d0f6d4ef460ce5f52f9b44fa712e7c81ad931c7e8a173a6d2c3132bb028980d6`

**Comment by youchengsong** [comment]
**[MEDIUM] Restrict free-form exemption to required choice questions**

Reviewer: `senior_engineer` · Category: `correctness`

Evidence: validate_input_answers computes one global has_free_form_answer flag and changes the required-question check to allow any missing required value whenever response_text is non-empty. Thus a request containing a required text/number question can be submitted unanswered merely by entering arbitrary Additional context. The frontend's missingRequired calculation has the same global exemption.

Impact: Plans can resume with required non-choice inputs missing, despite the feature requirement only allowing unanswered required choice options to be replaced by additional context. Downstream Planner/Reviewer calls may operate without mandatory data.

Required fix: Only allow the response_text exemption when the missing question is a supported choice response type; continue rejecting missing required text/number/etc. questions. Mirror that per-question rule in PlanInputForm's missingRequired calculation.

Verification: Add backend and frontend tests with one required single-choice question and one required text question: non-empty response_text with both structured values missing should return 422/keep Submit disabled, while a missing choice plus non-empty response_text should remain accepted.

Finding fingerprint: `d00a4bffc0de04ed1b64f0044f1ffcdabe43501b554e8ab165353e5d6c293cfc`

**Review by youchengsong — COMMENTED** [review]
# CCM reviewer panel: changes required

The panel found 3 open blocking findings on this reviewed PR head. Address the inline finding threads before merging.

## Reviewer summaries

### Principal engineer — Changes required

The revision-context design is cohesive, but the exhausted-feedback action is incorrectly exposed for historical versions.

Finding count: 1 total (1 blocking, 0 advisory).

### Senior engineer — Changes required

The revision-context changes are well tested, but the free-form choice escape hatch can silently bypass unrelated required choices.

Finding count: 1 total (1 blocking, 0 advisory).

### QA engineer — Changes required

Base-Version feedback is displayed in the prompt but is not wired into the mandatory feedback-closure path for a new revision Run.

Finding count: 1 total (1 blocking, 0 advisory).

Detailed evidence, impact, required fixes, and verification steps are published in dedicated finding threads/comments.

Reviewed commit: `dea6b323c46b0cdf45188bcd0d65f15109ef6353`.

**Comment by youchengsong** [comment]
**[MEDIUM] A new revision does not require closure of the Base Version's unresolved feedback**

Reviewer: `qa_engineer` · Category: `correctness`

Evidence: `previous_reviewer_feedback` is populated only from `run.review_feedback`. At the start of a new `user_revision` Run this is unset, while the selected Base Version's unresolved `base.review_feedback` appears only inside the generic `base_review_context` JSON. Consequently the explicit instruction to “verify every item before approving” sees `(none; perform a fresh complete review)`. The added integration test merely checks that the Base feedback text occurs somewhere in the prompt, so it does not detect this distinction.

Impact: A normal revision from an exhausted Version can be approved without explicitly checking or closing that Version's outstanding Reviewer findings, contrary to the PR's claimed revision semantics. The dedicated latest-feedback UI action embeds feedback in its request, but API callers and the ordinary revision action remain affected.

Required fix: For the first review cycle of a new Run, pass applicable unresolved Base Version feedback into the explicit feedback-to-verify/resolve field, falling back to it when `run.review_feedback` is absent. Preserve `run.review_feedback` precedence for subsequent same-Run cycles.

Verification: Extend the revision runner test so the first Planner and first Reviewer prompts place `Retain an explicit rollback procedure` under their actionable feedback sections, and verify a later same-Run Reviewer receives the newer `Specify cache rollback behavior` feedback instead.

Finding fingerprint: `998451f2936d56c6b630d6f40f9ceba94364765d5ba4528d2ca9304735f81c09`

File: backend/services/plan_service.py [hunk 1]
```diff
@@ -1248,21 +1248,42 @@ def _answer_map(answers: Iterable[PlanInputAnswer | dict]) -> dict[str, object]:
...
+    has_free_form_answer = bool(response_text and response_text.strip())
     unknown = set(values) - set(by_id)
     if unknown:
         raise HTTPException(422, f"Unknown question ids: {sorted(unknown)}")
     normalized: list[dict] = []
     for question in parsed:
         value = values.get(question.id)
-        if question.required and (value is None or value == "" or value == []):
-            raise HTTPException(422, f"Question {question.id!r} requires an answer")
+        free_form_replaces_choice = (
```
Comment (youchengsong) [inline backend/services/plan_service.py]: **[MEDIUM] One global response bypasses every unanswered required choice**

Reviewer: `senior_engineer` · Category: `correctness`

Evidence: `free_form_replaces_choice` depends only on the question being a choice and global `has_free_form_answer`. Therefore any non-empty `response_text` exempts all missing required choice questions. The frontend applies the same global rule without recording which question's “None of these options fit” action was selected. For two required choices, context answering only one—or unrelated optional context—allows both null answers through validation.

Impact: Runs can resume without required user decisions. The model is then instructed to treat each null choice plus the response as an explicit answer even when the response does not address that question, potentially producing an incorrect plan or consuming additional interaction rounds to recover.

Required fix: Record the user's free-form-replacement intent per question and exempt only those explicitly marked choice questions. The UI's alternative action should set that marker, and the backend must validate it against the corresponding question instead of using global response presence.

Verification: Submit two required choice questions with both values null, explicitly mark free-form replacement for only the first, and provide text addressing it; assert the API rejects the unanswered second question. Also verify unrelated additional context alone does not exempt an unmarked required choice.

Finding fingerprint: `3a7222c59478c7239862737b3d2ac51f908049ddb016d99b418c6c6919881a84`

**Comment by youchengsong** [comment]
**[MEDIUM] Historical exhausted versions incorrectly show the latest-feedback revision action**

Reviewer: `principal_engineer` · Category: `correctness`

Evidence: `canReviseWithLatestReviewerFeedback` checks that `shown` and `current` exist but never verifies `shown.id === plan.current_version_id`. Consequently, selecting an exhausted historical version can render the action. The click handler then silently returns because it separately enforces `shown.id === plan.current_version_id`. This also contradicts the added test that expects the button to remain absent after selecting historical version 11.

Impact: Users see an enabled revision action that does nothing, and the supplied PlanDetail test should fail for an exhausted historical version.

Required fix: Add the current-version identity check to `canReviseWithLatestReviewerFeedback`, preferably `shown.id === plan.current_version_id` or `shown.id === current.id`, so rendering and handler eligibility share the same invariant.

Verification: Run the added test `does not offer the latest-feedback revision for a Reviewer-approved Version` and add or retain an assertion that an exhausted historical version does not render the action while the current version differs.

Finding fingerprint: `80e22db873f798444b149fdaec2ffc80dada6e0c3d7e134e3f30ca3daac22008`

**Review by youchengsong — COMMENTED** [review]
# CCM reviewer panel: changes required

The panel found 3 open blocking findings on this reviewed PR head. Address the inline finding threads before merging.

## Reviewer summaries

### Principal engineer — Changes required

The new actions have two state-validation gaps: read-only Plans can expose a revision mutation, and optional free-form choices can produce a client-valid request that the backend rejects.

Finding count: 2 total (2 blocking, 0 advisory).

### Senior engineer — Changes required

The revision-context changes are coherent, but the free-form choice UI can submit a payload that backend validation necessarily rejects.

Finding count: 1 total (1 blocking, 0 advisory).

### QA engineer — Passed

The patch matches its stated revision-scope and free-form-choice behavior, adds focused backend and frontend coverage, and presents no concrete blocking regression in the supplied diff.

Finding count: none.

Detailed evidence, impact, required fixes, and verification steps are published in dedicated finding threads/comments.

Reviewed commit: `a4f6223272d7423469321fff76c3e4588c159c30`.

File: frontend/src/components/PlanReview/PlanInputForm.tsx [hunk 2]
```diff
@@ -41,9 +46,12 @@ export function PlanInputForm({ run, request, compact = false, onAnswered }: Pla
     () => request.questions.some((question) => {
       if (!question.required) return false;
```
Comment (youchengsong) [inline frontend/src/components/PlanReview/PlanInputForm.tsx:47]: **[MEDIUM] Optional free-form choices can leave the form in an invalid submit state**

Reviewer: `senior_engineer` · Category: `correctness`

Evidence: The alternative-answer button is rendered for every choice question and adds its ID to `freeFormQuestionIds`, while `missingRequired` examines only required questions. Thus, after marking an optional choice and leaving Additional context empty, the form may enable submission and send `answered_in_response_text: true`. `validate_input_answers` then rejects that payload because every marked answer requires a non-empty `response_text`. Clicking the already-pressed button also cannot remove the marker; only selecting an option clears it.

Impact: A user can follow an exposed UI action and receive an avoidable 422 response, and an optional question can effectively become mandatory after an accidental click.

Required fix: Keep submission disabled whenever any question is marked as answered in additional context but Additional context is empty, and allow the pressed action to toggle that marker off; alternatively, expose this action only for required choices if optional free-form answers are not supported.

Verification: Add a form test with no unanswered required questions and one optional choice: mark its additional-context action, verify submission remains disabled while context is blank, then verify either entering context enables submission or toggling the action off restores valid submission without `answered_in_response_text`.

Finding fingerprint: `52005b0da8e316600e97a06a91a22ad55b6738762bb6a41d78d667d19623503d`

File: frontend/src/components/PlanReview/PlanDetail.tsx [hunk 3]
```diff
@@ -328,6 +360,15 @@ export function PlanDetail({ plan, onRefresh, onClose, selectedVersionIds = [],
   const staleMessages = planStalenessMessages(staleness);
   const hardConflictMessages = planHardConflictMessages(staleness);
   const showStaleness = Boolean(shown && !shown.applied && !plan.read_only);
```
Comment (youchengsong) [inline frontend/src/components/PlanReview/PlanDetail.tsx:362]: **[MEDIUM] Hide the revision action for read-only Plans**

Reviewer: `principal_engineer` · Category: `regression`

Evidence: `canReviseWithLatestReviewerFeedback` checks current-version identity, active Run, archive state, decision, exhaustion, and feedback, but does not check `plan.read_only`. The adjacent `showStaleness` calculation explicitly honors `!plan.read_only`, and the supplied Plan contract describes migrated Plans as read-only history.

Impact: A read-only historical Plan can be offered a mutation that calls `createPlanRun`. This violates the aggregate's read-only state and either presents a predictably failing action or relies on downstream enforcement to prevent an invalid revision.

Required fix: Add `!plan.read_only` to `canReviseWithLatestReviewerFeedback`.

Verification: Render a current, pending, exhausted Version with feedback on a Plan whose `read_only` is true and assert that “Revise with Reviewer's latest feedback” is absent and `createPlanRun` is not called.

Finding fingerprint: `9999dd67903aa487940c280d3b76bb7dafa04c09b5f384d11242bcea5c4744f2`

File: frontend/src/components/PlanReview/PlanInputForm.tsx (hunk 2 above)
Comment (youchengsong) [inline frontend/src/components/PlanReview/PlanInputForm.tsx:47]: **[MEDIUM] Require additional text whenever a choice is marked as answered there**

Reviewer: `principal_engineer` · Category: `correctness`

Evidence: The alternative-answer button is rendered for optional as well as required choice questions and adds the question to `freeFormQuestionIds`. However, `missingRequired` examines only required questions. Submission can therefore remain enabled with an optional question marked `answered_in_response_text` and an empty `additional` value, while backend validation explicitly rejects every such marker without a non-empty `response_text`.

Impact: A valid UI flow submits a payload the API deterministically rejects with 422, leaving users unable to complete the form without discovering an unstated constraint.

Required fix: Treat any non-empty `freeFormQuestionIds` set with blank `additional` text as a client-side validation failure, and expose an explanatory message.

Verification: Use an optional choice question, activate “answer in additional context,” leave Additional context empty, and assert Submit remains disabled; then enter text and assert it becomes enabled and sends the marker with `response_text`.

Finding fingerprint: `943fbad87b2bcf30447b87ee4affd0d928622067d9e8388f3421fcc36950e1a1`

**Review by youchengsong — COMMENTED** [review]
# CCM reviewer panel: changes required

The panel found 1 open blocking finding on this reviewed PR head. Address the inline finding threads before merging.

## Reviewer summaries

### Principal engineer — Changes required

The revision and answer-state design is mostly coherent, but the UI does not expose the newly supported free-form path for multi-choice questions.

Finding count: 1 total (1 blocking, 0 advisory).

### Senior engineer — Passed

The revision-scope propagation, reviewer-feedback binding, explicit free-form choice handling, and UI state transitions are covered with focused validation and tests. No concrete blocking correctness, security, architecture, or maintainability issue was identified in the supplied patch.

Finding count: none.

### QA engineer — Passed

The revision-scope, reviewer-feedback, free-form choice, and exhausted-review flows are covered by focused backend and frontend tests, with no concrete blocking regression identified in the supplied patch.

Finding count: none.

Detailed evidence, impact, required fixes, and verification steps are published in dedicated finding threads/comments.

Reviewed commit: `3066b3de60c1294a2492bef7dc536ffbed87e646`.

**Comment by youchengsong** [comment]
**[MEDIUM] Multi-choice questions cannot use the free-form alternative**

Reviewer: `principal_engineer` · Category: `correctness`

Evidence: The backend validator explicitly treats both `single_choice` and `multi_choice` as eligible for `answered_in_response_text`, and the prompt describes choice questions generally. However, the new “None of these options fit” control remains inside a `question.response_type === 'single_choice'` conditional, so no such control or `freeFormQuestionIds` entry can be created for a multi-choice question.

Impact: A required multi-choice question whose generated options do not fit remains impossible to answer through the supported additional-context escape hatch; the user is forced to select an unsuitable option or cannot submit, contradicting the server contract and documented behavior.

Required fix: Render the free-form toggle for `single_choice` and `multi_choice` questions (or deliberately remove multi-choice from the backend/prompt contract), and keep the per-question marker behavior identical for both.

Verification: Add a PlanInputForm test with a required `multi_choice` question that marks it as answered in additional context, submits a null value with `answered_in_response_text: true`, and verifies `answerPlanInput` receives that payload; retain a test that another unmarked required question still blocks submission.

Finding fingerprint: `20c0481c899a3b02f3298a5e095f873e08bcfd1b0178967e2647412d0610f732`

**Comment by youchengsong** [comment]
CCM verified this Finding as fixed by the green exact-head review `12e51a183bf77376aadefebe013676b562ce82a5`.

**Comment by youchengsong** [comment]
CCM verified this Finding as fixed by the green exact-head review `12e51a183bf77376aadefebe013676b562ce82a5`.

**Comment by youchengsong** [comment]
CCM verified this Finding as fixed by the green exact-head review `12e51a183bf77376aadefebe013676b562ce82a5`.

**Comment by youchengsong** [comment]
CCM verified this Finding as fixed by the green exact-head review `12e51a183bf77376aadefebe013676b562ce82a5`.

**Comment by youchengsong** [comment]
CCM verified this Finding as fixed by the green exact-head review `12e51a183bf77376aadefebe013676b562ce82a5`.

**Comment by youchengsong** [comment]
CCM verified this Finding as fixed by the green exact-head review `12e51a183bf77376aadefebe013676b562ce82a5`.

**Review by youchengsong — COMMENTED** [review]
CCM automated review passed. The exact reviewed head `12e51a183bf77376aadefebe013676b562ce82a5` is ready to merge.
